sdk/storage/storage-blob/src/utils/StructuredMessageEncodingStream.ts

```diff
@@ -0,0 +1,461 @@
+/// without multi-reading the real stream, even when partway through an existing stream buffer.
+/// </item>
+/// <item>
+/// Else:
+/// Same as #1, but also the already-allocated stream buffer has been used to slightly improve
+/// resource churn when reading inner stream.
+/// </item>
+/// </list>
+/// </remarks>
+
+import { Readable } from "stream";
+import { AbortError } from "@azure/abort-controller";
+import { HttpRequestBody } from "../Pipeline";
+import { isNodeReadableStream } from "./utils.common";
+import { StorageCRC64Calculator } from "./StorageCRC64Calculator";
+import { isNodeLike } from "@azure/core-util";
+
+const MESSAGE_VERSION: number = 1;
+const MESSAGE_HEADER_LENGTH: number = 13;
+const SEGMENT_HEADER_LENGTH: number = 10;
+const FOOTER_LENGTH: number = 8;
+const MAX_SEGMENT_CONTENT_LENGTH = 4 * 1024 * 1024;
+
+enum SMRegion {
+  StreamHeader,
+  StreamFooter,
+  SegmentHeader,
+  SegmentFooter,
+  SegmentContent,
+  Completed
+}
+
+export interface StructuredMessageEncodingStreamOptions {
+  /**
+   * Read progress event handler
+   */
+  // TODO: onProgress?: (progress: TransferProgressEvent) => void;
+
+  /**
+   * Debug purpose only. Used to inject an unexpected end to existing internal stream,
+   * to test stream retry works well or not.
+   *
+   * When assign it to true, for next incoming "data" event of internal stream,
+   * RetriableReadableStream will try to emit an "end" event to existing internal
+   * stream to force it end and start retry from the breaking point.
+   * The value will then update to "undefined", once the injection works.
+   */
+  doInjectErrorOnce?: boolean;
+
+  /**
+   * A threshold, not a limit. Dictates the amount of data that a stream buffers before it stops asking for more data.
+   */
+  highWaterMark?: number;
+}
+
+async function pump(
+  reader: ReadableStreamDefaultReader, 
+  controller: ReadableStreamDefaultController,
+  encodingStream: StructuredMessageEncoding) : Promise<void> {
+  const { done, value } = await reader.read();
+  // When no more data needs to be consumed, close the stream
+  if (done) {
+    controller.close();
+    return;
+  }
+  
+  // Enqueue the next data chunk into our target stream
+  encodingStream.sourceDataHandler(value);
+}
+
+export async function BrowserStream(source: Blob | ReadableStream<Uint8Array>, content_length: number): Promise<{content: Blob, encodedContentLength: number}> {
+  const sourceStream = (source instanceof Blob) ? source.stream() : source;
+  const reader = sourceStream.getReader();
+
+  let encodingStream : StructuredMessageEncoding | undefined = undefined;
+  const stream = new ReadableStream({
+    start(controller) { 
+      encodingStream = new StructuredMessageEncoding((data) => {
+        controller.enqueue(data)
+      }, content_length);
+    },
+    pull (controller) {
+      pump(reader, controller, encodingStream!).then(() =>{});
+    }
+  });
+
+  const response = new Response(stream);
```

Comment:
For uploading in Browser, seems I'd need to read everything in a Blob (in memory) before uploading. Is there anyway to read&handle data(calculate checksum and append after the content) while uploading?